File: backend/app/routes/cortes.py
"""
Rotas /cortes — CRUD de planos de redução de gastos.
Cada corte representa uma economia mensal planejada ou aprovada.
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from app.services.storage_service import cortes_storage

router = APIRouter()


class CorteCreate(BaseModel):
    descricao: str
    categoria: str
    economia_mensal: float
    status: str = "planejado"   # planejado | em_analise | aprovado | aplicado
    ativo: bool = True
    observacao: Optional[str] = None


class CorteUpdate(BaseModel):
    descricao: Optional[str] = None
    categoria: Optional[str] = None
    economia_mensal: Optional[float] = None
    status: Optional[str] = None
    ativo: Optional[bool] = None
    observacao: Optional[str] = None


@router.get("")
def list_cortes():
    """Lista todos os cortes com impacto acumulado."""
    cortes = cortes_storage.all()
    ativos = [c for c in cortes if c.get("ativo", True)]

    economia_total  = sum(c.get("economia_mensal", 0) for c in cortes)
    economia_ativa  = sum(c.get("economia_mensal", 0) for c in ativos)

    return {
        "cortes": cortes,
        "resumo": {
            "total":                     len(cortes),
            "ativos":                    len(ativos),
            "economia_total_potencial":  round(economia_total, 2),
            "economia_ativa":            round(economia_ativa, 2),
            "impacto_3m":                round(economia_ativa * 3, 2),
            "impacto_6m":                round(economia_ativa * 6, 2),
            "impacto_12m":               round(economia_ativa * 12, 2),
        },
    }


@router.post("")
def create_corte(body: CorteCreate):
    """Cria um novo item de corte/redução."""
    return cortes_storage.create(body.model_dump())


@router.put("/{id}")
def update_corte(id: str, body: CorteUpdate):
    """Atualiza um corte existente (toggle ativo/inativo incluído)."""
    payload = body.model_dump()
    # Remove None, mas mantém ativo=False (truthy check não funciona aqui)
    update_data = {k: v for k, v in payload.items() if v is not None or k == "ativo"}
    update_data = {k: v for k, v in update_data.items() if v is not None}
    atualizado = cortes_storage.update(id, update_data)
    if not atualizado:
        raise HTTPException(status_code=404, detail="Corte não encontrado")
    return atualizado


@router.delete("/{id}")
def delete_corte(id: str):
    """Remove um corte pelo id."""
    if not cortes_storage.delete(id):
        raise HTTPException(status_code=404, detail="Corte não encontrado")
    return {"ok": True}

File: functions/src/routes/cortes.ts
import { Router } from "express";
import { cortesStorage } from "../lib/firestore";

const router = Router();

function toFloat(v: unknown): number {
  const n = parseFloat(String(v ?? "0"));
  return isFinite(n) ? n : 0;
}

// GET /cortes
router.get("/", async (_req, res) => {
  try {
    const items = await cortesStorage.all();
    const ativos = items.filter((c) => c.status === "ativo");
    const total = items.length;
    const economia_ativa = ativos.reduce((s, c) => s + toFloat(c.valor), 0);
    const r = Math.round;
    res.json({
      cortes: items,
      resumo: {
        total,
        ativos: ativos.length,
        economia_total_potencial: r(items.reduce((s, c) => s + toFloat(c.valor), 0) * 100) / 100,
        economia_ativa: r(economia_ativa * 100) / 100,
        impacto_3m: r(economia_ativa * 3 * 100) / 100,
        impacto_6m: r(economia_ativa * 6 * 100) / 100,
        impacto_12m: r(economia_ativa * 12 * 100) / 100,
      },
    });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// POST /cortes
router.post("/", async (req, res) => {
  try {
    res.status(201).json(await cortesStorage.create(req.body));
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// PUT /cortes/:id
router.put("/:id", async (req, res) => {
  try {
    const updated = await cortesStorage.update(req.params.id, req.body);
    if (!updated) { res.status(404).json({ detail: "Corte não encontrado" }); return; }
    res.json(updated);
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// DELETE /cortes/:id
router.delete("/:id", async (req, res) => {
  try {
    if (!await cortesStorage.delete(req.params.id)) { res.status(404).json({ detail: "Corte não encontrado" }); return; }
    res.json({ ok: true });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

export default router;
